### PR TITLE
[training] Tensorflow interface for MultiNode SGD

### PR DESCRIPTION
**Pull request by jichan3751**
## What do these changes do?
Creates Tensorflow interface for MultiNode SGD.

TODO:

 - [x] Smoke test data_augmentation_creator
 - [x] Smoke test regular data_creator
 - [x] Write docs
 - [x] Verify that tests pass
 - [x] Verify that this works on multi-node multi-gpu

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by richardliaw** [comment]
BTW, I don't think you need to fit the PyTorch API so closely. I think you should first get the Distributed TF example running in Ray, and then think about APIs afterwards.

**Comment by jichan3751** [comment]
Looks like there are problems with model save / loading.
There are basically 3 data needed to restore a model:
  1. sturucture config
  2. weights
  3. optimizer weights
These are saved/ loaded without problems when writing / reading from disk using model.save().
However, it is little tricky to apply optimizer weights if we want to restore the model from the python objects: optimizer weights = [] if model.fit() is not called. and this is not letting us setting the weight from python weight object. 
I included lots of hacks to get around this. Let me know your suggestion.

File: python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import os
+import json
+import tensorflow as tf
+
+from ray.experimental.sgd.tensorflow.tensorflow_runner import TensorFlowRunner
+
+logger = logging.getLogger(__name__)
+
+
+class DistributedTensorFlowRunner(TensorFlowRunner):
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py]: Is the inheritance necessary here?

Comment (jichan3751) [reply python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py]: class TensorFlowRunner's method get_state and set_state can be used to get current model. So I think inheritance is needed here.

File: python/ray/experimental/sgd/examples/tensorflow_train_example.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+from ray import tune
+from ray.experimental.sgd.tensorflow.tensorflow_trainer import (
+    TensorFlowTrainer, TensorFlowTrainable)
+
+from ray.experimental.sgd.tests.tf_helper import (get_model, get_dataset)
```
Comment (richardliaw) [inline python/ray/experimental/sgd/examples/tensorflow_train_example.py]: For this, I think we should not use `tf_helper` and define `get_model` and `get_dataset` in this file. PyTorch should be updated in a separate PR.

File: python/ray/experimental/sgd/examples/tf-example-sgd.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    #         MarketType: spot
+
+setup_commands:
+    - conda install setuptools=41.0.1=py36_0 wrapt=1.11.2 --yes # workaround to fix wrapt error
+    - ray || pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev3-cp36-cp36m-manylinux1_x86_64.whl
+    - pip install -U ipdb ray[rllib]
+    - pip install -U tensorflow-gpu==2.0.0-beta1
+
+file_mounts: {
+    ~/run/: /Users/jichanchung/OneDrive/FF/190812_tf2_tune_trainable/190814_multinode_mnist_ray
```
Comment (richardliaw) [inline python/ray/experimental/sgd/examples/tf-example-sgd.yaml]: remove?

File: python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py]: I think you should just define `get_state` and `set_state` in this class, so I think we don't need a separate DistributedTensorFlowRunner clas - just a regular TensorFlowRunner.

File: python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            # model.fit() returns None for MultiWorkerMirroredStrategy
+            stats = {}
+        else:
+            stats = {"train_loss": history.history["loss"][-1]}
+
+        self.epoch += 1
+
+        return stats
+
+    def validate(self):
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tensorflow/distributed_tensorflow_runner.py]: Validate should only be called on the `test_dataset`.

File: python/ray/experimental/sgd/tensorflow/tensorflow_runner.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+from __future__ import absolute_import
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tensorflow/tensorflow_runner.py]: You should merge this with DistributedTensorflowRunner, as commented in that file

File: python/ray/experimental/sgd/tensorflow/tensorflow_runner.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        return {
+            "epoch": self.epoch,
+            "weights": self.model.get_weights(),
+            "optimizer_weights": self.model.optimizer.get_weights()
+        }
+
+    def set_state(self, state):
+        self.epoch = state["epoch"]
+        if self.model.optimizer.weights == []:
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tensorflow/tensorflow_runner.py]: I don't quite understand this issue nor this comment - can you:

1. provide a comment of exactly what the error is, and
2. provide in the code a link to the stackoverflow or Tensorflow github issue link that you found which suggested this workaround?

File: python/ray/experimental/sgd/tests/tf_helper.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tests/tf_helper.py]: this should be separate lines, and I don't think you need unicode_laterals

File: python/ray/experimental/sgd/tests/tf_helper.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import tensorflow as tf
+
+NUM_TRAIN_SAMPLES = 60000
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tests/tf_helper.py]: can we make this 512?

File: python/ray/experimental/sgd/tests/tf_helper.py (hunk 7 above)
Comment (richardliaw) [inline python/ray/experimental/sgd/tests/tf_helper.py]: can you copy these functions to the example?

File: python/ray/experimental/sgd/utils.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+from __future__ import division
+from __future__ import print_function
+
+from contextlib import closing
+import numpy as np
+import socket
+import time
+
+
+class TimerStat(object):
```
Comment (richardliaw) [inline python/ray/experimental/sgd/utils.py]: should this be removed from the Torch utils?

File: python/ray/experimental/sgd/tests/tf_helper.py (hunk 8 above)
Comment (jichan3751) [reply python/ray/experimental/sgd/tests/tf_helper.py]: shuffle(NUM_TRAIN_SAMPLES) is used to shuffle whole data. 
Are you meaning we should only consider 512 datapoints from mnist dataset?

File: python/ray/experimental/sgd/utils.py (hunk 9 above)
Comment (jichan3751) [reply python/ray/experimental/sgd/utils.py]: I think it is better to reflect in separate branch.

**Comment by jichan3751** [comment]
Do you know proper way make pytest recognize sgd/tensorflow directory?
when I run 'pytest tests', it says 'ImportError: cannot import name 'TensorFlowTrainer' from 'ray.experimental.sgd.tensorflow.tensorflow_trainer''
while pytest test_tensorflow.py is running fine if run in the tests/ directory.

File: python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+        return checkpoint
+
+    def restore(self, checkpoint):
+        """Restores the model from the provided checkpoint.
+
+        Args:
+            checkpoint (str): Path to target checkpoint file.
+
+        """
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py]: we don’t need state[config] right? and you dont need to sabe the entire model right (no need for model.save) right? You can simply use model_creator to achieve the same effect.

File: python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+        state = ray.get(self.workers[0].get_state.remote())
+
+        model = self._get_model_from_state(state)
+
+        model.save(checkpoint + ".h5")
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py]: like technically we just need to dump the “state” to file via pickle; no need to use model. save

File: python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py (hunk 10 above)
Comment (jichan3751) [reply python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py]: Yes for the state['config'], I'll remove it

**Review by richardliaw — COMMENTED** [review]
this looks really good! just one last comment about serialization. I’m going to try it out on aws tonight. Don’t forget to kill your instances.

**Comment by jichan3751** [comment]
Thanks!
Could you give me comment about this issue?

> Do you know proper way make pytest recognize sgd/tensorflow directory?
> when I run 'pytest tests', it says 'ImportError: cannot import name 'TensorFlowTrainer' from 'ray.experimental.sgd.tensorflow.tensorflow_trainer'' in tensorflow/**init**.py
> while pytest test_tensorflow.py is running fine if run in the tests/ directory.

**Comment by richardliaw** [comment]
Hm, I’m not quite sure... I’ll look into it tonight.

File: python/ray/experimental/sgd/utils.py (hunk 9 above)
Comment (jichan3751) [reply python/ray/experimental/sgd/utils.py]: (moved)

File: python/ray/experimental/sgd/examples/cifar-tf-example.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    batch_size = args.batch_size
+
+    model = train_example(
+        args.num_replicas,
+        batch_size=batch_size,
+        epochs=args.num_epochs,
+        use_gpu=args.use_gpu)
+    dataset, test_dataset = data_augmentation_creator(batch_size=batch_size)
+    model.fit(dataset, steps_per_epoch=60000 // batch_size, epochs=1)
+    scores = model.evaluate(test_dataset, steps=10000 // batch_size)
```
Comment (richardliaw) [inline python/ray/experimental/sgd/examples/cifar-tf-example.py]: TODO: test both data_augmentation and normal versions. Set smoke test, which only takes 10 steps.

File: python/ray/experimental/sgd/examples/train_example.py [hunk 13]
```diff
@@ -12,6 +12,10 @@
 from __future__ import division
 from __future__ import print_function
 
+import torch
+import torch.nn as nn
+import numpy as np
```
Comment (richardliaw) [inline python/ray/experimental/sgd/examples/train_example.py]: todo: remove

**Review by richardliaw — APPROVED** [review]
